cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -154,59 +155,48 @@ public void updateAppSpec(String namespaceId, String appId, ApplicationSpecifica
-    if (record == null) {
-      String msg = String.format("No meta found for associated Workflow %s run record %s, while recording run for the" +
-                                   " namespace %s app %s type %s program %s runid %s", workflowRunId.getProgram(),
-                                 workflowRunId.getRun(), program.getNamespaceId(), program.getApplicationId(),
-                                 program.getType().name(), program.getId(), pid);
-      LOG.error(msg);
-      throw new IllegalArgumentException(msg);
+    Map<String, WorkflowNodeStateDetail> nodeStates = new HashMap<>();
+    for (WorkflowNodeStateDetail nodeStateDetail : nodeStateList) {
+      nodeStates.put(nodeStateDetail.getNodeId(), nodeStateDetail);
     }
+    return nodeStates;
+  }
 
-    // Update the parent Workflow run record by adding node id and program run id in the properties
-    Map<String, String> properties = record.getProperties();
-    properties.put(workflowNodeId, pid);
-
-    RunRecordMeta runRecordMeta = new RunRecordMeta(record, properties);
-    String failureMsg = failureCause == null ? null : Throwables.getRootCause(failureCause).getMessage();
-    WorkflowNodeState nodeState = new WorkflowNodeState(workflowNodeId, ProgramRunStatus.toNodeStatus(status),
-                                                        pid, failureMsg);
+  public void addWorkflowNodeState(ProgramRunId workflowRunId, WorkflowNodeStateDetail nodeStateDetail) {
+    // Node states will be stored with following key:
+    // workflowNodeState.namespace.app.WORKFLOW.workflowName.workflowRun.workflowNodeId
+    MDSKey key = getProgramKeyBuilder(TYPE_WORKFLOW_NODE_STATE, workflowRunId.getParent().toId())
+      .add(workflowRunId.getRun()).add(nodeStateDetail.getNodeId()).build();
 
-    RunRecordMeta meta = updateRunRecordWithWorkflowNodeState(runRecordMeta, nodeState);
-    write(key, meta);
+    write(key, nodeStateDetail);
   }
 
-  private RunRecordMeta updateRunRecordWithWorkflowNodeState(RunRecordMeta runRecordMeta, WorkflowNodeState nodeState) {
-    Map<String, WorkflowNodeState> nodeStates;
-    String workflowNodeStates = runRecordMeta.getProperties().get(WORKFLOW_NODE_STATE_KEY);
-    if (workflowNodeStates == null) {
-      nodeStates = new HashMap<>();
-    } else {
-      nodeStates = GSON.fromJson(workflowNodeStates, new TypeToken<Map<String, WorkflowNodeState>>() { }.getType());
-    }
+  private void addWorkflowNodeState(Id.Program program, String pid, Map<String, String> systemArgs,
+                                    ProgramRunStatus status, @Nullable Throwable failureCause) {
+    String workflowNodeId = systemArgs.get(ProgramOptionConstants.WORKFLOW_NODE_ID);
+    String workflowName = systemArgs.get(ProgramOptionConstants.WORKFLOW_NAME);
+    String workflowRun = systemArgs.get(ProgramOptionConstants.WORKFLOW_RUN_ID);
+
+    ProgramRunId workflowRunId = new ProgramRunId(program.getNamespaceId(), program.getApplicationId(),
```

Comment:
you should get this from the caller of this method, but this can be ``programId.run(workflowRun)``